### PR TITLE
Replace sq_rsp(...) with a new DSYEV wrapper, part 1

### PR DESCRIPTION
**Pull request by TiborGY**
## Description
`sq_rsp(...)` is an ancient diagonalizer function from Psi3 with many issues. While its innards could be improved, the function signature is horrid. It should be deprecated and replaced with something better.
Stemming from the awful interface, users of `sq_rsp(...)` never check if the diagonalization failed, because they _cannot_. This is now rectified by adding failure checks wherever the new DSYEV wrappers are called.
Checks against non-square matrices are also added where there is both a row and column count at the call site.

Some of the call sites would be dfocc territory, this PR does not include them to avoid conflicts with the dfocc overhaul.
This is another shard of the https://github.com/psi4/psi4/pull/2642 mega-PR that can be merged after PR #2678 and #2738 are in.

## Todos

- [x] Implement two new wrapper functions around DSYEV to replace `sq_rsp(...)` with. The new wrappers have a much cleaner interface, and no longer require the allocation of an eigenvector array if the caller only needs eigenvalues. They also no longer swallow the return value of DSYEV, in fact their return value is marked as `[[nodiscard]]`. In short, checking for diagonalization failure went from impossible to mandatory.
Please note that unlike `sq_rsp(...)`, the new functions do not take a "tolarence" value, but this is actually a null change as that parameter has been set-but-unused in `sq_rsp(...)` since ~forever ago when `sq_rsp(...)` switched to calling DSYEV internally.
- [x] Replace `sq_rsp(...)` calls in `mcscf` (`mcscf::SCF::energy` and `mcscf::MatrixBase::diagonalize`) with new wrapper. Add checks for diagonalization failure and guard against non-square matrices.
- [x] Replace `sq_rsp(...)` calls in `detci` (`detci/h0block.cc`, `detci/sem.cc` and `detci/sem_test.cc`) with new wrapper. Add checks for diagonalization failure.
- [x] Replace `sq_rsp(...)` call in `libmints/matrix.cc` (`Matrix::diagonalize`) with new wrapper. Add checks for diagonalization failure. Add a sanity checks for non-square matrices and illegal values of "diagonalization order".
- [x] Replace `sq_rsp(...)` calls in `ccenergy` (`ccenergy/d1diag.cc`, `ccenergy/d2diag.cc` and `ccenergy/new_d1diag.cc`) with new wrapper. Add checks for diagonalization failure. This allows the removal of the eigenvector array, and the code that allocates/deallocates it.
- [x] Replace `sq_rsp(...)` calls in the `libqt` Davidson solver with new wrapper. Add checks for diagonalization failure.
- [x] Replace `sq_rsp(...)` call in the RHF and ROHF stability checks with new wrapper. Add checks for diagonalization failure.
- [x] Replace `sq_rsp(...)` calls in `libsapt_solver/sapt2.cc` with new wrapper. Add checks for diagonalization failure.

## Questions
- [x] Is `libciomr` a good place for the new wrappers to live in? The one they are replacing is there, so it seemed like as good of a place as any.

## Checklist
- [x] No new features
- [x] 503/503 ctests are passing

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by psi-rking** [comment]
As an old-timer, I would say yes, libciomr is the place.  At least as long as there is still a libciomr, that's where I would look.  This is nice improvement!

ccdensity has a sq_rsp too.  did you miss that one?

**Comment by TiborGY** [comment]
> As an old-timer, I would say yes, libciomr is the place. At least as long as there is still a libciomr, that's where I would look. This is nice improvement!
> 
> ccdensity has a sq_rsp too. did you miss that one?

I ran a search and none of the remaining hits were from ccdensity.

**Comment by psi-rking** [comment]
in cc/ccdensity/ael.cc
This code determines approximate excitation level.  It used to work.
Apparently no longer called.  I'm not sure why.
(I'm not on the very latest psi4, so it's possible it's been excised.

**Comment by TiborGY** [comment]
Indeed it has, by yours truly. See #2672 for the removal. I do not know when the AEL code became defunct, but I removed it because it was never called. The code is preserved in the psi4attic repo.

File: psi4/src/psi4/libciomr/dsyev_ascending.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+** \param e_vecs = (optional) matrix of eigenvectors (2D row major array, one column for each eigvector)
+**
+** \ingroup CIOMR
+*/
+[[nodiscard]] int DSYEV_ascending(const int N, const double* const* const array, double* e_vals,
+                                  double* const* const e_vecs /* = nullptr*/) {
+    // We need to make a copy of the matrix before diagonalization, because LAPACK overwrites it.
+    // LAPACK also needs the mtx to be flattened to a 1D array, so a copy is inevitable.
+    // The new 1D array will correspond to a column-major array, suitable for LAPACK.
+    double* tmp_matrix = init_array(N * N);
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libciomr/dsyev_ascending.cc:54]: Use `std::vector<double> tmp_matrix(N * N)` instead. Any function signature expecting a `double *` will receive `tmp_matrix.data()` instead. This frees us from needing to manage memory manually.

Comment (TiborGY) [reply psi4/src/psi4/libciomr/dsyev_ascending.cc:54]: Good idea!

**Review by loriab — APPROVED** [review]
Nice cleanup and consolidation, thank you!

**Comment by TiborGY** [comment]
503/503 ctest tests are passing
[ctestlog.txt](https://github.com/psi4/psi4/files/10106557/ctestlog.txt)
If that is judged to be sufficient coverage, this PR is ready for merge.

**Comment by loriab** [comment]
I've just run full ctests and pytests with this branch, and it's clean, so lgtm!

**Review by JonathonMisiewicz — APPROVED** [review]

**Pull request merged**